lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -294,6 +304,113 @@ struct PathTracker {
   SetVector<StringAttr> altBasePathRoots;
 };
 
+/// Constants and helpers for creating the RtlPorts on the fly.
+
+static constexpr StringRef kContainingModuleName = "containingModule";
+static constexpr StringRef kPortsName = "ports";
+static constexpr StringRef kRtlPortClassName = "RtlPort";
+
+static Type getRtlPortsType(MLIRContext *context) {
+  return om::ListType::get(om::ClassType::get(
+      context, FlatSymbolRefAttr::get(context, kRtlPortClassName)));
+}
+
+static om::ListCreateOp
+createRtlPorts(firrtl::PathOp containingModuleRef, Value basePath,
+               const PathInfoTable &pathInfoTable,
+               hw::InnerSymbolNamespaceCollection &namespaces,
+               HierPathCache &hierPathCache, OpBuilder &builder,
+               std::mutex &intraPassMutex) {
+  // Look up the module from the containingModuleRef.
+
+  FlatSymbolRefAttr containingModulePathRef =
+      pathInfoTable.table.at(containingModuleRef.getTarget()).symRef;
+
+  const SymbolTable &symbolTable = hierPathCache.getSymbolTable();
+
+  hw::HierPathOp containingModulePath =
+      symbolTable.lookup<hw::HierPathOp>(containingModulePathRef.getAttr());
+
+  assert(containingModulePath.isModule() &&
+         "expected containing module path to target a module");
+
+  StringAttr moduleName = containingModulePath.leafMod();
+
+  FModuleLike mod = symbolTable.lookup<FModuleLike>(moduleName);
+  MLIRContext *ctx = mod.getContext();
+  Location loc = mod.getLoc();
+
+  // Create the per-port information.
+
+  auto portClassName = StringAttr::get(ctx, kRtlPortClassName);
+  auto portClassType =
+      om::ClassType::get(ctx, FlatSymbolRefAttr::get(portClassName));
+
+  SmallVector<Value> ports;
+  for (unsigned i = 0, e = mod.getNumPorts(); i < e; ++i) {
+    // Only process ports that are not zero-width.
+    auto portType = type_dyn_cast<FIRRTLBaseType>(mod.getPortType(i));
+    if (!portType || portType.getBitWidthOrSentinel() == 0)
+      continue;
+
+    // Get a path to the port. This may modify port attributes or the global
+    // namespace of hierpaths, so use the mutex around those operations.
+
+    auto portTarget = PortAnnoTarget(mod, i);
+
+    FlatSymbolRefAttr portPathRef;
+    {
+      std::lock_guard<std::mutex> guard(intraPassMutex);
+
+      auto portSym =
+          getInnerRefTo({portTarget.getPortNo(), portTarget.getOp(), 0},
+                        [&](FModuleLike m) -> hw::InnerSymbolNamespace & {
+                          return namespaces[m];
+                        });
+
+      portPathRef = hierPathCache.getRefFor(ArrayAttr::get(ctx, {portSym}));
```

Comment:
`getRefFor` could create HierPathOp in a top-level module so the order could be non-deterministic due to multithreading.

Reply:
Hmm yes good point. I put this section under a mutex to try to make sure it was thread safe, but I hadn't thought about non-deterministic order. I don't think it would matter in practice, but I guess in general we should strive for deterministic MLIR outputs.

I'm trying to think about how to make this deterministic... I guess when we are doing updateInstances in parallel, we can collect objects that need this extra field and process them later single-threaded. But that would still need to be smart about handling the objects that were collected from multiple threads in a stable order. Maybe I could come up with a stable sort for the collected objects and then process them serially in that order?

Reply:
I took a crack at this in [962f859](https://github.com/llvm/circt/pull/7651/commits/962f8598e7a0ee8d059a9f19619626354a519e96). Doing the multi-threaded collection and then single-threaded post processing also helped clean up the overall diff, I think it's simpler this way.

Regarding the actual determinism, I sorted the collected objects by their class name before processing single threaded. There could still be non-determinism if we collected objects of the same class from multiple threads, but in practice this should never happen; we only add this field to a handful of objects and they should be unique within each circuit.